### ISSUE TITLE
How does the implementation achieve layer-by-layer training?

### ISSUE THREAD

**Issue by matthiasgeihs** [loeweX/Forward-Forward#3]
Hey, for someone not super familiar with how PyTorch autograd works internally:
Would you be able to shed some light on how your implementation realizes the two forward passes and layer by layer training?

(It seems like you do everything in one pass (positive, negative, layer-by-layer training), while in other implementations it is more obvious how they perform layer by layer optimization.)

**Comment by loeweX** [comment]
Hey,

In PyTorch, one can stop the flow of gradients by calling the detach() function on tensors. For example, you can see this happening in [this line](https://github.com/loeweX/Forward-Forward/blob/92c0c1c0565063dae0121734075013cab20da5c5/src/ff_model.py#L113). This ensures that each layer is only trained based on the gradients from its own local loss. 

Regarding the separation of the positive and negative samples: running the positive samples first, then the negative samples and adding their gradients together (as done in the original MatLab implementation) will result in the same weight update as concatenating the positive and negative samples into a single batch and calculating their gradients together.

Hope this helps!

**Comment by qqqllppp** [comment]
Hi @loeweX , I am a bit confused on this implementation of the ff algo. Reading the paper, it seems that it should replace the backprop part altogether, but from this [line](https://github.com/loeweX/Forward-Forward/blob/da516f094c70a4a92042fbbc7ab33303aa7dbfd9/main.py#L26C1-L26C46), it seems that the weights of each layer are only updated after calling the backward() function Also, the detact() function is only called on `z`, so layer weights are still considered by backward(). So it seems this implementation still uses backprop to learn/update weights?

I may have missed something, do correct me if so :)
